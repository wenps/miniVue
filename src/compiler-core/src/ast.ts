export const enum NodeTypes {
    INTERPOLATION = 'interpolation', // 插值类型，如 {{message}}
    SIMPLE_EXPRESSION = 'simple_expression', // 简单表达式，如 xiaoshan
    ELEMENT = 'element', // 元素类型
    TEXT = 'text', // 文本类型
    ROOT = 'root', // 根节点
}