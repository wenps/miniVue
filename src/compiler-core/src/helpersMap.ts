// 需要加载的依赖
export const TO_DISPLAY_STRING = Symbol('toDisplayString')

export const helperMapName = {
    [TO_DISPLAY_STRING]: 'toDisplayString'
}