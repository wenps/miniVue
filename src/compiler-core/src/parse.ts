import { NodeTypes } from "./ast"
enum TagType  {
    start,
    end
}

export function baseParse(content:string) {

    // 创建一个全局上下文对象，包裹传进来的 content
    const context = createParserContext(content)

    return createRoot(parseChildren(context))
}

// 封装content逻辑
function createParserContext(content: string) {
    return {
     source:content
    }
 }

// 抽离根节点逻辑
function createRoot(children) {
    return {
        children
    }
}

// 抽离children节点逻辑
function parseChildren(context) {
    const nodes:any = [] // 创建节点列表
    let node;
    const s = context.source
    // 判断是否插值类型节点
    if(s.startsWith('{{')) {
        node = parseInterpolation(context) // 解析插值类型
    } 
    // 判断是否元素类型节点，是否以 < 开头，第二个字段是否a-z中的字符
    else if (s[0] === '<') {
        if(/[a-z]/i.test(s[1])) {
            node = parseElement(context) // 解析元素类型
        }
    }
    nodes.push(node) // 设置节点列表
    return nodes
}

// 抽离插值类型节点逻辑
function parseInterpolation(context) {

    const openDelimiter = '{{' 
    const closeDelimiter = '}}'

    // 下述操作相当于 获取 {{message}} 中的 message，截取插值类型内容
    const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length) // 拿到 }} 下标
    advanceBy(context, openDelimiter.length) // 拿到 message}} ， 推进两位 {{
    const rawContentLength = closeIndex - openDelimiter.length 
    const rawContent = context.source.slice(0, rawContentLength) // 拿到 message, 即原始数据
    const content = rawContent.trim()
    advanceBy(context, rawContentLength + closeDelimiter.length) // 推进 message.length + 2 位 （上面推进了{{ 和 message.length + 2 位 这样子 {{message}}整体被推进完了）
    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION, 
            content: content
        }
    }
}

function parseElement(context) {
    const element = parseTag(context, TagType.start) // 处理开头标签
    parseTag(context, TagType.end) // 处理结束标签
    return element
    
}

function parseTag(context: any, type: TagType) {
    // 1.解析tag
    const match: any = /^<\/?([a-z]*)/i.exec(context.source) // 标签解析正则
    const tag = match[1]
    // 2.处理后的代码推进
    advanceBy(context, match[0].length + 1) // 推进掉<div + >

    if(type == TagType.start) { // 如果是开头标签就返回
        return {
            type: NodeTypes.ELEMENT,
            tag,
        }
    }
}

// 封装推进代码
function advanceBy(context:any, length:number) {
    context.source = context.source.slice(length)
}
