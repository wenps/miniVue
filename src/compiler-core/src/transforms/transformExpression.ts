import { NodeTypes } from "../ast";

// 设置插值类型下，普通表达式的 render函数格式 , 这里是加了_ctx.
export function transformExpression(node) {
    if(node.type == NodeTypes.INTERPOLATION) {
        const rawContent = node.content.content
        node.content.content = "_ctx." + rawContent
    }
}