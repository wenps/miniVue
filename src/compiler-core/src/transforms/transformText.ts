import { NodeTypes } from '../ast';

// 创建复合类型，将相邻text和插值类型整合起来
export function transformText(node) {
    // 判断当前是否是text或插值
    function isText(node) {
        return node.type == NodeTypes.TEXT || node.type == NodeTypes.INTERPOLATION;
    }

    if (node.type == NodeTypes.ELEMENT) {
        return () => {
            // 获取当前元素节点的children
            const { children } = node;

            // 创建一个容器存储复合类型
            let currentContainer;

            // 遍历children
            for (let i = 0; i < children.length; i++) {
                const child = children[i];

                if (isText(child)) {
                    // 判断当前是否是text或插值，如果是开启循环判断下一个是不是，找到第一个不是的为止，再把上述的内容加到新的复合类型节点下
                    for (let j = i + 1; j < children.length; j++) {
                        const nextChild = children[j];
                        if (isText(nextChild)) {
                            // 如果是text或者插值

                            // 如果容器不存在，默认初始化一个容器，并替换掉当前节点child
                            if (!currentContainer) {
                                currentContainer = children[i] = {
                                    // 替换掉当前节点
                                    type: NodeTypes.COMPOUND_EXPRESSION,
                                    children: [child]
                                };
                            }

                            currentContainer.children.push(' + ');
                            currentContainer.children.push(nextChild); // 将符合条件的节点加入到容器中

                            // 并且删除数组中已经被加入的节点，这样子这次循环结束结束之后，下一个必然是非text或插值类型的node
                            children.splice(j, 1);
                            // 因为删了一个之后j会指向下一个，有因为长度变了就会出现这种情况。
                            // 假设 1 2 3 4  此时 j 在 2 的 位置 即下标 1，我们 把 2 加进去又删掉 2，j ++  j就等下标 2
                            // 因为 2 没了 下标2 直接到 了4的位置，越过了 3 这显然是不正确的，因此加进去的节点 需要再 j--
                            j--;
                        } else {
                            // 非 text或者插值
                            currentContainer = null; // 容器重置，跳出小循环
                            break;
                        }
                    }
                }
            }
        };
    }
}
