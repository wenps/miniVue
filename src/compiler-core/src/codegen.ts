import { NodeTypes } from "./ast"
import { TO_DISPLAY_STRING, helperMapName } from "./helpersMap"

export function generate(ast) {

    const context = createCodegenContext()
    const {push} = context
    
    // 组装成render函数

    // 抽离导入资源类型代码
    genFunctionPreamble(push, ast)

    // render函数返回部分组装
    push('return')

    const functionName = "render"
    const args = ['_ctx', '_cache']
    const signature = args.join(",")

    push(` function ${functionName}(${signature}){`)
    push('return ')

    genNode(ast.codegenNode, push)
    push('}')
    

    return {
        code: context.code 
    }
}

// render 核心节点 加入
function genNode(node: any, push) {
    // 基于ast 完善render函数主体内容
    switch (node.type) {
        case NodeTypes.TEXT: // 处理text类型
            genText(node, push)
            break;

        case NodeTypes.INTERPOLATION:  // 处理插值类型
            genInterpolation(node, push)
            break;

        case NodeTypes.SIMPLE_EXPRESSION:  // 处理简单表达式类型
            genSimpleExpression(node, push)
            break;
    
        default:
            break;
    }
    
    
}

// 创建一个code相关的全局上下文对象
function createCodegenContext() {
    const context = {
        code: '',
        push(source) {
            context.code += source
        },
    }
     return context
}

// 抽离导入资源类型代码
function genFunctionPreamble(push: (source: any) => void, ast: any) {
    
    // helperMapName 依赖资源映射表
    const vue = '"vue"'
    
    const aliasHelpers = (s) => `${helperMapName[s]} as _${helperMapName[s]}`
    
    // 判断当前是否需要导入资源，插值类型需要，但是text类型是不用的，因此这里判断一下
    if (ast.helpers.length > 0) {
        push(`import { ${ast.helpers.map(aliasHelpers).join(", ")} } from ${vue} `)
        push('\n')
    }

}

// render 核心节点 加入 text类型处理
function genText(node: any, push: any) {
    push(`"${node.content}"`)
}

// render 核心节点 加入 插值类型处理
function genInterpolation(node: any, push: any) {
    push(`_${helperMapName[TO_DISPLAY_STRING]}(`)
    genNode(node.content, push)
    push(')')
}

// render 核心节点 加入 简单表达式类型处理
function genSimpleExpression(node: any, push: any) {
    push(`${node.content}`)
}

