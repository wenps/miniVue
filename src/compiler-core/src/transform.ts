import { NodeTypes } from "./ast"
import { TO_DISPLAY_STRING, helperMapName } from "./helpersMap"

export function transform(root, options = {}) {

    const context = createTransformContext(root, options) // 封装一个全局上下文对象

    // 遍历 - 深度优先搜索
    traverseNode(root, context)

    // root.codegenNode 基于此创建render代码
    createRootCodegen(root)

    root.helpers = [...context.helpers.keys()]
}

// 为根节点提供一个指向默认编译节点的属性
function createRootCodegen(root: any) {
    root.codegenNode = root.children[0]
}

// 遍历 - 深度优先搜索
function traverseNode(node, context) {

    const nodeTransforms = context.nodeTransforms // 获取全局上下文中的插件数组
    

    // 遍历执行插件数组
    for (let i = 0; i < nodeTransforms.length; i++) {
        nodeTransforms[i](node);
    }

    // 针对不同类型节点作操作
    switch (node.type) {
         // 插值类型
        case NodeTypes.INTERPOLATION:
            context.helper(TO_DISPLAY_STRING) // 往全局对象上加上依赖，因为解析插值类型需要导入这个依赖
            break;
         // 元素类型和根节点类型
        case NodeTypes.ROOT:
        case NodeTypes.ELEMENT:
            // 递归子节点
            traverseChildren(node, context) // 因为元素和根节点都有children ，所有需要递归children
            break;
        default:
            break;
    }

    
}

// 递归子节点
function traverseChildren(node, context) {
    const children = node.children
    if (children) {
        for (let i = 0; i < children.length; i++) {
            const curNode = children[i];
            traverseNode(curNode, context) // 递归
        }
    }
}

// 封装一个全局上下文对象
function createTransformContext(root: any, options: any) {
    
    const context = {
        root,
        nodeTransforms: options.nodeTransforms || [],
        // 封装需要加载的依赖
        helpers: new Map(),
        helper(key) {
            context.helpers.set(key, 1)
        }
    }
    return context
}

