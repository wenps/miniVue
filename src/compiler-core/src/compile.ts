import { generate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"
import { transformElement } from "../src/transforms/transformElement"
import { transformExpression } from "../src/transforms/transformExpression"
import { transformText } from "../src/transforms/transformText"
export function baseCompile(template) {
    const ast: any = baseParse(template)

        transform(ast, {
            nodeTransforms: [transformExpression, transformElement, transformText] // 加上一个插件，处理元素类型的依赖
        })
        
        

        return generate(ast)

        // console.log(code);
}