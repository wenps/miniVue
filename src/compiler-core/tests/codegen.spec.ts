import { generate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"
import { transformExpression } from "../src/transforms/transformExpression"

describe('codegen', ()=>{
    it('happy pass',()=>{
        const ast = baseParse('hi')

        transform(ast)

        const { code } = generate(ast)

        // 快照测试
        expect(code).toMatchSnapshot()
    })
    it('happy pass',()=>{
        const ast = baseParse('{{message}}')

        transform(ast, {
            nodeTransforms: [transformExpression] // 加上一个插件，设置插值类型下，普通表达式的 render函数格式 , 这里是加了_ctx.
        })

        console.log(ast);
        

        const { code } = generate(ast)

        // 快照测试
        expect(code).toMatchSnapshot()
    })
})