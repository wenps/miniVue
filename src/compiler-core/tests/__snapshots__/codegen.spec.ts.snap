// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen happy pass 1`] = `"return function render(_ctx,_cache){return "hi"}"`;

exports[`codegen happy pass 2`] = `
"import { toDisplayString as _toDisplayString } from "vue" 
return function render(_ctx,_cache){return _toDisplayString(_ctx.message)}"
`;
